torch/quantization/_adaround.py

```diff
@@ -0,0 +1,260 @@
+
+def clipped_sigmoid(continous_V):
+    ''' Function to create a non-vanishing gradient for V
+
+    Paper Reference: https://arxiv.org/pdf/2004.10568.pdf Eq. 23
+    '''
+    sigmoid_of_V = torch.sigmoid(continous_V)
+    scale_and_add = (sigmoid_of_V * 1.2) - 0.1
+    return torch.clamp(scale_and_add, 0, 1)
+
+def adaround_rounding(model, x):
+    ''' Using the scale and continous_V parameters of the model, the given tensor x is
+    rounded using adaround
+
+    Paper Reference: https://arxiv.org/pdf/2004.10568.pdf Eq. 22
+    '''
+    weight = x
+    continous_V = model.continous_V
+    scale = model.scale
+
+    weights_divided_by_scale = torch.div(weight, scale)
+    weights_divided_by_scale = torch.floor(weights_divided_by_scale)
+    weights_clipped = weights_divided_by_scale + clipped_sigmoid(continous_V)
+
+    weights_w_adaround_rounding = scale * torch.clamp(weights_clipped, model.quant_min, model.quant_max)
+    return weights_w_adaround_rounding
+
+def loss_function_leaf(model, count):
+    ''' Calculates the loss function for a submodule
+
+    Paper Reference: https://arxiv.org/pdf/2004.10568.pdf Eq. 25
+    '''
+    beta = count / number_of_epochs * (beta_high - beta_low) + beta_low
+
+
+    # Calculates the difference between floating point and quantized models
+    adaround_instance = model.wrapped_module.weight_fake_quant
+    float_weight = model.wrapped_module.weight
+    clipped_weight = adaround_rounding(adaround_instance, float_weight)
+    quantized_weight = torch.fake_quantize_per_tensor_affine(clipped_weight, float(adaround_instance.scale),
+                                                             int(adaround_instance.zero_point), adaround_instance.quant_min,
+                                                             adaround_instance.quant_max)
+    # Frobenius_norm = torch.norm(float_weight - quantized_weight)  # norm(W - ~W)
+    Frobenius_norm = torch.norm(model.float_output - model.quantized_output)  # norm(Wx - ~Wx)
+
+    # calculating regularization factor -> forces values of continous_V to be 0 or 1
+    scale = adaround_instance.scale
+    continous_V = adaround_instance.continous_V
+    clip_V = clipped_sigmoid(continous_V)
+    spreading_range = torch.abs((2 * clip_V) - 1)
+    one_minus_beta = 1 - (spreading_range ** beta)  # torch.exp
+    regulization = torch.sum(one_minus_beta)
+
+    print("loss function break down: ", Frobenius_norm * norm_scaling, regularization_scaling * regulization)
+    # print("sqnr of float and quantized: ", computeSqnr(float_weight, quantized_weight))
+    return Frobenius_norm * norm_scaling + regularization_scaling * regulization
+
+def loss_function(model, count):
+    ''' Given model, the loss function of all of its whitelisted leaf modules will
+    be added up and returned.
+    '''
+    result = torch.Tensor([0])
+    for name, submodule in model.named_modules():
+        if isinstance(submodule, OuputWrapper):
```

Comment:
do we need this wrapper? can you just check the type of the module directly? if it is AdaRoundFakeQuantize module then we will accumulate the loss